src/Forms/GridFieldQueuedExportButton.php

```diff
@@ -0,0 +1,314 @@
+        $job->setIncludeHeader($this->csvHasHeader);
+        if ($this->exportColumns) {
+            $job->setColumns($this->exportColumns);
+        }
+
+        // Queue the job
+        singleton(QueuedJobService::class)->queueJob($job);
+
+        // Redirect to the status update page
+        return Controller::curr()->redirect($gridField->Link('/export/' . $job->getSignature()));
+    }
+
+
+    /**
+     * This class is also a URL handler
+     *
+     * @param GridField $gridField
+     * @return array
+     */
+    public function getURLHandlers($gridField)
+    {
+        return [
+            'export/$ID' => 'checkExport',
+            'export_download/$ID' => 'downloadExport'
+        ];
+    }
+
+    protected function getExportPath($id)
+    {
+        if ($id instanceof QueuedJobDescriptor) {
+            $id = $id->Signature;
+        }
+        return ASSETS_PATH . "/.exports/$id/$id.csv";
+    }
+
+
+    /**
+     * Handle the export, for both the action button and the URL
+     *
+     * @param GridField $gridField
+     * @param HTTPRequest $request
+     * @return HTTPResponse|DBHTMLText
+     */
+    public function checkExport($gridField, $request = null)
+    {
+        $id = $request->param('ID');
+        $job = QueuedJobDescriptor::get()->filter('Signature', $id)->first();
+
+        if ((int)$job->RunAsID !== Security::getCurrentUser()->ID) {
+            return Security::permissionFailure();
+        }
+
+        $controller = $gridField->getForm()->getController();
+
+        $breadcrumbs = $controller->Breadcrumbs(false);
+        $breadcrumbs->push(new ArrayData([
+            'Title' => _t('SilverStripe\\Forms\\GridField\\GridField.CSVEXPORT', 'Export to CSV'),
+            'Link' => false
+        ]));
+
+        $parents = $controller->Breadcrumbs(false)->items;
+        $backlink = array_pop($parents)->Link;
+
+        $data = new ArrayData([
+            'ID' => $id,
+            'Link' => Controller::join_links($gridField->Link(), 'export', $job->Signature),
+            'Backlink' => $backlink,
+            'Breadcrumbs' => $breadcrumbs,
+            'GridName' => $gridField->getname()
+        ]);
+
+        if ($job->JobStatus == QueuedJob::STATUS_COMPLETE) {
+            if (file_exists($this->getExportPath($id))) {
+                $data->DownloadLink = $gridField->Link('/export_download/' . $job->Signature);
+            } else {
+                $data->ErrorMessage = _t(
+                    'SilverStripe\\GridfieldQueuedExport\\GridFieldQueuedExportButton.ERROR_REMOVED',
+                    'This export has already been downloaded. ' .
+                    'For security reasons each export can only be downloaded once.'
+                );
+            }
+        } elseif ($job->JobStatus == QueuedJob::STATUS_BROKEN) {
+            $data->ErrorMessage = _t(
+                'SilverStripe\\GridfieldQueuedExport\\GridFieldQueuedExportButton.ERROR_GENERAL',
```

Comment:
This one actually is _this_ class, so use of the constant `__CLASS__` is probably better here :)

Reply:
Thanks, I changed it.